server/src/main/java/io/druid/server/lookup/cache/LookupCoordinatorManager.java

```diff
@@ -506,137 +274,426 @@ public boolean deleteLookup(final String tier, final String lookup, AuditInfo au
    *
    * @return The lookupName spec if found or null if not found or if no lookups at all are found
    */
-  public
   @Nullable
-  Map<String, Object> getLookup(final String tier, final String lookupName)
+  public LookupExtractorFactoryMapContainer getLookup(final String tier, final String lookupName)
   {
-    final Map<String, Map<String, Map<String, Object>>> prior = getKnownLookups();
+    final Map<String, Map<String, LookupExtractorFactoryMapContainer>> prior = getKnownLookups();
     if (prior == null) {
       LOG.warn("Requested tier [%s] lookupName [%s]. But no lookups exist!", tier, lookupName);
       return null;
     }
-    final Map<String, Map<String, Object>> tierLookups = prior.get(tier);
+    final Map<String, LookupExtractorFactoryMapContainer> tierLookups = prior.get(tier);
     if (tierLookups == null) {
       LOG.warn("Tier [%s] does not exist", tier);
       return null;
     }
     return tierLookups.get(lookupName);
   }
 
-
-  @LifecycleStart
+  // start() and stop() are synchronized so that they never run in parallel in case of ZK acting funny and
```

Comment:
well this code is being called from DruidCoordinator (coordinator leadership code) and not from TaskMaster. You may be right that start/stop here are never called concurrently but I just want to be sure and I don't think overhead of synchronization matters here.

LifecycleLock canStart/canStop are not equivalent of synchronization. For example , consider following sequence of events.
1. start() is called and goes through lifecycleLock.canStart() succesfully.
2. before start() is finished , now stop() is called which fails at lifecycleLock.canStop()

in this case, caller intended the end state to be stopped, while it will be left in started state.